tiatoolbox/tools/graph.py

```diff
@@ -0,0 +1,275 @@
+from sklearn.neighbors import KDTree as sKDTree
+
+
+def hybrid_clustered_graph(
+    points: ArrayLike,
+    features: ArrayLike,
+    label: Optional[int] = None,
+    lambda_d: Number = 3.0e-3,
+    lambda_f: Number = 1.0e-3,
+    lambda_h: Number = 0.8,
+    connectivity_distance: Number = 4000,
+    neighbour_search_radius: Number = 2000,
+    feature_range_thresh: Number = 1e-4,
+) -> Dict[str, ArrayLike]:
+    """Build a graph via hybrid clustering in spatial and feature space.
+
+    This is part of the SlideGraph pipeline but may be used to construct
+    a graph in general from point coordinates and features.
+
+    The graph is constructed via hybrid heirachical clustering and
+    Delaunay triangulation.
+
+    The clustering uses a distance kernel, ranging between 0 and 1,
+    which is a weighted product of spatial distance and feature-space
+    distance. Points which are spatially further apart than
+    `neighbour_search_radius` are given a similarity of 1 (most
+    dissimilar). This significantly speeds up computation. This distance
+    metric is then used to form clusters via hierachical/agglomerative
+    clustering.
+
+    Next, a Delaunay triangulation is applied to the clusters to connect
+    the neighouring clusters.
+
+    Args:
+        points (ArrayLike): A list of (x, y) coordinates within a WSI.
+        features (ArrayLike): A list of features associated with each
+            coordinate in `points`. Must be the same length as `points`.
+        lambda_d (Number): Spatial distance weighting.
+        lambda_f (Number): Feature space distance weighting.
+        lambda_h (Number): Clustering distance threshold. Applied to
+            the similarity kernel (1-fd). Ranges between 0 and 1.
+            Defaults to 0.8. A good value for this parameter will depend
+            on the intra-cluster variance.
+        connectivity_distance (Number):
+            Distance threshold to consider points as connected during
+            during the Delaunay triangulation step.
+        neighbour_search_radius (Number):
+            Search radius (L2 norm) threshold for points to be
+            considered as connected. Points with a spatial distance
+            above this are not compared and have a similarity set to 1
+            (most dissimilar).
+        feature_range_thresh (Number):
+            Minimal range for which a feature is considered significant.
+            Features which have a range less than this are ignored.
+            Defaults to 1e-4.
+
+    Returns:
+        dict: A dictionary defining a graph for serialisation (e.g.
+        JSON) or converting into a torch-geometric Data object.
+            - :class:`numpy.ndarray` - x:
+                Features of each node (mean of features in a cluster).
+            - :class:`numpy.ndarray` - edge_index:
+                Edge index matrix defining connectivity.
+            - :py:obj:`Number` - y:
+                The label of the graph.
```

Comment:
Return a set of nodes where each is a 2D spatial cluster from the input points?

The return/example description doesn't describe well what each value means, and the initial intro also does not.